src/commands/up.rs

```diff
@@ -646,6 +668,76 @@ fn trigger_command_for_resolved_app_source(
         .collect()
 }
 
+// Scrubs the locked app to only contain the given list of components
+// Introspects the LockedApp to find and selectively retain the triggers that correspond to those components
+fn retain_components(locked_app: &mut LockedApp, component_trigger_id_map: ComponentTriggerMap) {
+    let components = component_trigger_id_map.keys().collect::<HashSet<_>>();
+    let triggers = component_trigger_id_map.values().collect::<HashSet<_>>();
+    locked_app.components.retain(|c| components.contains(&c.id));
+    locked_app.triggers.retain(|t| triggers.contains(&t.id));
+}
+
+// Given a list of components and App, returns a map of component ID to trigger ID.
+// Also validates that the components exist in the app and have valid service chaining.
+fn component_trigger_map(
+    app: &spin_app::App,
+    components: &[String],
+) -> Result<ComponentTriggerMap> {
+    let component_triggers: Vec<(AppComponent<'_>, String)> = app
+        .triggers()
+        .filter_map(|t| match t.component() {
+            Ok(comp) => {
+                if components.contains(&comp.id().to_string()) {
+                    Some((comp, t.id().to_string()))
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        })
+        .collect();
+
+    // Allow only service chaining among retained components
+    // All wildcard service chaining is disallowed
+    for (c, _) in &component_triggers {
+        let allowed_hosts = allowed_hosts(c)?;
+        allowed_hosts.iter().try_for_each(|host| {
+            let uri = host.parse::<http::Uri>().unwrap();
+            if let Some(local_component) = parse_service_chaining_target(&uri) {
+                if !components.contains(&local_component) {
+                    if local_component == "*" {
+                        bail!("Wildcard service chaining not allowed in retained components");
```

Comment:
I think this could be wordsmithed to be a little more clear; maybe something like:

```
Component selected with '--component {id}' cannot use wildcard service chaining:
  allowed_outbound_hosts = ["http://*.spin.internal"]
```

Reply:
That message seems less clear to me: it is no longer clear to me that the 'selection' is the _cause_ of not being able to use the wildcard service chaining.  I read it as "this component, which is included in the selection, can't use wildcard service chaining."

Reply:
Yeah its definitely a little awkward; alternative wordsmithing welcome!

Reply:
I think the problem with the original wording is that "wildcard service chaining" and "retained components" are not necessarily familiar concepts.

Reply:
>  I read it as "this component, which is included in the selection, can't use wildcard service chaining."

I think this is an accurate reading. No retained component is allowed to use wildcard service chaining. They can only use precise/targeted service chaining

Reply:
@kate-goldenring Ironically, I poorly worded my comment about the wording.  What I read was that "this component can't use wildcard service chaining (for some reason)" (and Spin needs to tell you about this because this component is selected).  Your original wording made it clear that the problem is that wildcard service chaining isn't compatible with selection.

How about something like:

"You cannot use the `*.spin.internal` wildcard host in combination with the `--component` flag. Change the `allowed_outbound_hosts` for component 'isometric-darwin' to list specific components for service chaining, or remove the `--component` flag."

(This is wordy for sure, and may be too specific to CLI syntax, I'm just offering it as a sighting shot.)